File: P/915-partition-array-into-disjoint-intervals/main.go
package main

func partitionDisjoint(nums []int) int {
	left, right := make([]int, len(nums)), make([]int, len(nums))

	leftMax, rightMin := nums[0], nums[len(nums)-1]
	for i := 0; i < len(nums); i++ {
		if nums[i] > leftMax {
			leftMax = nums[i]
		}
		if nums[len(nums)-1-i] < rightMin {
			rightMin = nums[len(nums)-1-i]
		}
		left[i] = leftMax
		right[len(nums)-1-i] = rightMin
	}
	for i := 0; i < len(nums)-1; i++ {
		if left[i] <= right[i+1] {
			return i + 1
		}
	}
	return -1
}

File: P/915-partition-array-into-disjoint-intervals/main.rs
struct Solution {}

impl Solution {
  pub fn partition_disjoint(nums: Vec<i32>) -> i32 {
    let mut left: Vec<i32> = vec![0; nums.len()];
    let mut right: Vec<i32> = vec![0; nums.len()];

    let mut leftMax: i32 = nums[0];
    let mut rightMin: i32 = nums[nums.len() - 1];
    for i in (0..nums.len()) {
      if nums[i] > leftMax {
        leftMax = nums[i]
      }
      if nums[nums.len() - 1 - i] < rightMin {
        rightMin = nums[nums.len() - 1 - i]
      }
      left[i] = leftMax;
      right[nums.len() - 1 - i] = rightMin;
    }

    let mut ret: i32 = -1;
    for i in (0..nums.len() - 1) {
      if left[i] <= right[i + 1] {
        ret = (i + 1) as i32;
        break;
      }
    }
    ret
  }
}
